File: ruby/raindrops/raindrops.rb
FACTOR_MAPPINGS = [
  {factor: 3, output: 'Pling'},
  {factor: 5, output: 'Plang'},
  {factor: 7, output: 'Plong'},
]

class Raindrops
  def self.convert(num)
    factors = FactorFinder.find_factors(num)

    output = FACTOR_MAPPINGS.reduce('') { |memo, mapping| self.potentially_add_to_output memo, mapping, factors }

    if output.empty?
      num.to_s
    else
      output
    end
  end

  def self.potentially_add_to_output(output, mapping, factors)
    if factors.include? mapping[:factor]
      output += mapping[:output]
    else
      output
    end
  end

end

class FactorFinder
  def self.find_factors(num)
    [1, num] | (2..(num/2)).select { |i| (num % i).zero? }
  end
end

File: elixir/raindrops/raindrops.exs
defmodule Raindrops do

  @factor_pairs [
    "Pling": 3,
    "Plang": 5,
    "Plong": 7,
  ]

  @default ""

  @doc """
  Returns a string based on raindrop factors.

  - If the number contains 3 as a prime factor, output 'Pling'.
  - If the number contains 5 as a prime factor, output 'Plang'.
  - If the number contains 7 as a prime factor, output 'Plong'.
  - If the number does not contain 3, 5, or 7 as a prime factor,
    just pass the number's digits straight through.
  """
  @spec convert(pos_integer) :: String.t
  def convert(number) do
    factors_string = string_for_factors(number)
    if factors_string == @default do
      to_string number
    else
      factors_string
    end
  end

  @spec string_for_factors(pos_integer) :: String.t
  defp string_for_factors(number) do
    @factor_pairs
    |> Enum.reduce(@default, fn
      ({term, n}, str) when number >= n and rem(number, n) == 0 ->
        str <> to_string(term)
      (_, str) ->
        str
    end)
  end
end
